code/dirs.py

```diff
@@ -260,3 +307,230 @@ def makedirs_if_not_exist(directory):
+        roaming_appdir = os.path.join(home, "AppData", "Roaming")
+    
+        if (os.path.is_dir(local_appdir) and os.path.is_dir(roaming_appdir))
+            return (roaming_appdir, local_appdir)
+
+    # We can check the old ApplicationData
+    for home in home_dirs:
+        appdir = os.path.join(home, "Application Data")
+    
+        if (os.path.is_dir(appdir))
+            return (appdir, appdir)
+
+    # Ok, let just use the user directory.
+    for home in home_dirs:
+        if (os.path.is_dir(home))
+            return (home, home)
+
+    # Now, next we could be check the registry.
+    # But it's not good idea.
+    # 1) Registry shell32 are considered as unreliable
+    # 2) If we can't use win32 for shell32, why could we use it for registry ?
+
+    return False
+
+def get_mac_library():
+    
+    try:
+        import ctypes
+        import ctypes.util
+    except:
+        if g.debug: sys.stderr.write("WARNING: Impossible to load ctypes modules\n")
+
+    if 'ctypes' in sys.modules:
+        try:
+            NSApplicationSupportDirectory = 14
+            NSUserDomainMask = 1
+            
+            Foundation = ctypes.cdll.LoadLibrary("Foundation.framework/Foundation")
+            CoreFoundation = ctypes.cdll.LoadLibrary("CoreFoundation.framework/CoreFoundation");
+
+            NSSearchPathForDirectoriesInDomains = Foundation.NSSearchPathForDirectoriesInDomains
+            CFRelease = CoreFoundation.CFRelease
+            CFArrayGetCount = CoreFoundation.CFArrayGetCount
+            CFArrayGetValueAtIndex = CoreFoundation.CFArrayGetValueAtIndex
+            CFStringGetLength = CoreFoundation.CFStringGetLength
+            CFStringGetMaximumSizeForEncoding = CoreFoundation.CFStringGetMaximumSizeForEncoding
+            CFStringGetCString = CoreFoundation.CFStringGetCString
+
+            kCFStringEncodingUTF8 = 0x08000100
+
+            result = False
+
+            paths = NSSearchPathForDirectoriesInDomains(directory, domainMask, expand)
+
+            if paths and CFArrayGetCount(paths) > 0:
+                # TODO: Use all search paths found.
+
+                path = CFArrayGetValueAtIndex(paths, 0)
+
+                length = CFStringGetLength(path);
+                maxSize = CFStringGetMaximumSizeForEncoding(length, kCFStringEncodingUTF8) + 1;
+
+                buf = create_string_buffer(maxSize)
+                if CFStringGetCString(path, buf, sizeof(buf), kCFStringEncodingUTF8):
+                    result = buf.raw.decode('utf-8').rstrip('\0')
+                del buf
+
+            CFRelease(paths)
+
+            if result: return result
+            if g.debug: sys.stderr.write("WARNING: '%s' fondation function fails for library\n" % func_name)
+        except:
+            if g.debug: sys.stderr.write("WARNING: Impossible to use '%s' fondation function\n" % func_name)
+
+    # Only fallback is hardcoded dir in HOME
+    if os.environ.has_key("HOME"):
```

Comment:
Another case of `X.has_key(Y)` that should be `Y in X` (or rewritten as a try-except KeyError)